torsiondrive/qm_engine.py

```diff
@@ -327,6 +335,166 @@ def load_native_output(self, filename='output.dat'):
+                        # self.optblockStr will be set by self.optimize_native()
+                        outfile.write(self.optblockStr)
+
+                else:
+                    outfile.write(line)
+
+    def optimize_native(self):
+        """
+        Run the constrained optimization, following Gaussian09 manual.
+        1. write a optimization job input file.
+        2. run the job
+        """
+        assert self.temp_type == 'optimize', "To use native optimization, the input file be an opt job"
+        if self.extra_constraints is not None:
+            raise RuntimeError('extra constraints not supported in Gaussian09 native optimizations')
+        self.optblockStr =''
+        for d1, d2, d3, d4, v in self.dihedral_idx_values:
+            self.optblockStr += f'{d1 + 1} {d2 + 1} {d3 + 1} {d4 + 1} ={v:.3f} B\n'  # Build the angle
+            self.optblockStr += f'{d1 + 1} {d2 + 1} {d3 + 1} {d4 + 1} F\n'           # Freeze the angle
+        # write input file
+        self.write_input('gaussian.com')
+        # run the job
+        self.run('g09 < gaussian.com > gaussian.log', input_files=['gaussian.com'], output_files=['gaussian.log'])
+        self.run('formchk lig.chk lig.fchk')    # More reliable to get the geometry from the log file and convert it
+
+    def load_native_output(self, filename='lig.fchk', filename2='gaussian.log'):
+        """ Load the optimized geometry and energy into a new molecule object and return """
+        found_opt_result = False
+        final_energy, elems, coords = None, [], []
+        with open(filename2) as logfile:
+            logfile = logfile.readlines()
+            for counter, line in enumerate(logfile):
+                line = line.strip()
+                if line.startswith('Optimization completed'):
+                    found_opt_result = True
+
+        if found_opt_result is not True:
+            raise RuntimeError("Geometry optimisation failed in %s" % filename2)
+
+        with open(filename) as outfile:
+            outfile = outfile.readlines()
+            for counter, line in enumerate(outfile):
+                if line.startswith('Current cartesian coordinates'):
+                    start_xyz_pos = int(counter + 1)
+                    num_xyz=int(line.split()[5])
+                    end_xyz_pos=int(np.ceil(num_xyz/5)+start_xyz_pos)
+                if line.startswith('Total Energy'):
+                    energy_pos = counter
+
+        if not start_xyz_pos and end_xyz_pos:
+            raise EOFError('Cannot locate coordinates in lig.fchk file.')
+
+        for line in outfile[start_xyz_pos: end_xyz_pos]:
+            coords.extend([float(num) * 0.529177 for num in line.strip('\n').split()])
+
+        # Make sure we have all of the coordinates
+        assert len(coords) == num_xyz, "Could not extract the optimised geometry"
+
+        final_energy = float(outfile[energy_pos].split()[3])
+
+        if final_energy is None:
```

Comment:
I didn't find a check for the convergence of the constrained optimization. I guess what will happen here is that if the constrained optimization job failed to converge, this function will parse the output file and return the result from the last iteration as it converged successfully.

In the sample of `EnginePsi4`, you can see that the `load_native_output()` function looks for the string `"Final optimized geometry and variables"`, which is only printed when it successfully convergences. We should raise an Error when the optimization does not converge.

Reply:
```        if opt_result is not True:
            raise RuntimeError("Geometry optimization failed in %s" % filename2)```
The check is done just after reading through the log file so we don't extract the coords from the second file if this is not in the log file.